logfire/_internal/stack_info.py

```diff
@@ -65,3 +70,35 @@ def get_caller_stack_info(stack_offset: int = 3) -> StackInfo:
         return get_stack_info_from_frame(frame)
     except Exception:  # pragma: no cover
         return {}
+
+
+def get_user_stack_offset() -> int:
+    """Get the stack offset of the user code.
+
+    We want to skip the internal code, and third party code, and get the user code stack info.
+
+    Returns:
+        The stack offset of the user code.
+    """
+    stack_offset = 2
+    try:
+        frame = inspect.currentframe()
+        while frame and not is_user_filename(frame.f_code.co_filename):
+            frame = frame.f_back
+            stack_offset += 1
+    except Exception:  # pragma: no cover
+        pass
+    return stack_offset
+
+
+def is_user_filename(filename: str) -> bool:
+    """Check if the filename is a user filename.
+
+    Args:
+        filename: The filename to check.
+
+    Returns:
+        True if the filename is a user filename, False otherwise.
+    """
+    filename = get_filepath(filename)
+    return not (filename.startswith('logfire') or 'site-packages' in filename)
```

Comment:
Does this conditional make sense?

Reply:
I would prefer something like this:

```python
from pathlib import Path

import opentelemetry.sdk.trace

import logfire

site_packages_dir = str(Path(opentelemetry.sdk.trace.__file__).parent.parent.parent.parent.absolute())
logfire_dir = str(Path(logfire.__file__).parent.absolute())
prefixes = (site_packages_dir, logfire_dir)


def is_ignored(path):
    return str(Path(path).absolute()).startswith(prefixes)
```

Plus some generous caching - caching `get_code_object_info` was a big win because pathlib is slow.